cupyx/scipy/sparse/linalg/_solve.py

```diff
@@ -76,6 +78,286 @@ def lsqr(A, b):
+                    according to atol and btol.
+
+                    2 means x approximately solves the least-squares problem
+                    according to atol.
+
+                    3 means COND(A) seems to be greater than CONLIM.
+
+                    4 is the same as 1 with atol = btol = eps (machine
+                    precision)
+
+                    5 is the same as 2 with atol = eps.
+
+                    6 is the same as 3 with CONLIM = 1/eps.
+
+                    7 means ITN reached maxiter before the other stopping
+                    conditions were satisfied.
+
+            - itn (int): Number of iterations used.
+            - normr(float): ``norm(b-Ax)``
+            - normar (float): ``norm(A^T (b - Ax))``
+            - norma (float): ``norm(A)``
+            - conda (float): Condition number of A.
+            - normx (float): ``norm(x)``
+
+    .. seealso:: :func:`scipy.sparse.linalg.lsmr`
+
+    References:
+        D. C.-L. Fong and M. A. Saunders, "LSMR: An iterative algorithm for
+        sparse least-squares problems", SIAM J. Sci. Comput.,
+        vol. 33, pp. 2950-2971, 2011.
+    """
+    A = _interface.aslinearoperator(A)
+    b = b.squeeze()
+
+    m, n = A.shape
+    minDim = min([m, n])
+
+    if maxiter is None:
+        maxiter = minDim
+
+    u = b
+    beta = cublas.nrm2(b)
+
+    v = cupy.zeros(n)
+    alpha = 0
+
+    if beta > 0:
+        u = (1 / beta) * u
+        v = A.rmatvec(u)
+        alpha = cublas.nrm2(v)
+
+    if alpha > 0:
+        v = (1 / alpha) * v
+
+    # Initialize variables for 1st iteration.
+
+    itn = 0
+    zetabar = alpha * beta
+    alphabar = alpha
+    rho = 1
+    rhobar = 1
+    cbar = 1
+    sbar = 0
+
+    h = v.copy()
+    hbar = cupy.zeros(n)
+    x = cupy.zeros(n)
+
+    # Initialize variables for estimation of ||r||.
+
+    betadd = beta
+    betad = 0
+    rhodold = 1
+    tautildeold = 0
+    thetatilde = 0
+    zeta = 0
+    d = 0
+
+    # Initialize variables for estimation of ||A|| and cond(A)
+
+    normA2 = alpha * alpha
+    maxrbar = 0
+    minrbar = 1e+100
+    normA = alpha
+    condA = 1
+    normx = 0
+
+    # Items for use in stopping rules.
+    normb = beta
+    istop = 0
+    ctol = 0
+    if conlim > 0:
+        ctol = 1 / conlim
+    normr = beta
+
+    # Golub-Kahan process terminates when either alpha or beta is zero.
+    # Reverse the order here from the original matlab code because
+    # there was an error on return when arnorm==0
+    normar = alpha * beta
+    if normar == 0:
+        return x, istop, itn, normr, normar, normA, condA, normx
+
+    # Main iteration loop.
+    while itn < maxiter:
+        itn = itn + 1
+
+        # Perform the next step of the bidiagonalization to obtain the
+        # next  beta, u, alpha, v.  These satisfy the relations
+        #         beta*u  =  a*v   -  alpha*u,
+        #        alpha*v  =  A'*u  -  beta*v.
+
+        u = A.matvec(v) - alpha * u
```

Comment:
Would you use in-place oparators, i.g. `*=`, where we can? In-place operators on ndarray invoke in-place kernels and overwrite the contents of the left-hand side ndarray instance. On the other hand, Binary operators, i.g. `*`, on ndarray make a new ndarray instance for the result and allocate GPU memory from the memory pool.